src/psfmachine/machine.py

```diff
@@ -837,289 +635,197 @@ def build_time_model(self, plot=False, downsample=False, split_time_model=False)
+                mpc2_smooth.max() - mpc2_smooth.mean()
+            )
+            # combine them as the first order
+            other_vectors = [mpc1_smooth, mpc2_smooth, mpc1_smooth * mpc2_smooth]
+
+        # create a 3D perturbation matrix
+        P3 = PerturbationMatrix3D(
+            time=_whitened_time,
+            dx=dx,
+            dy=dy,
+            segments=split_time_segments,
+            focus=focus_component,
+            resolution=self.n_time_points,
+            other_vectors=other_vectors,
+            bin_method=bin_method,
+        )
 
-            # we solve the segment by using `seg_mask` on all `*_binned` variables
-            for count in [0, 1, 2]:
-                sigma_w_inv = A3[k].T.dot(
-                    (
-                        A3.multiply(1 / flux_err_binned[seg_mask].ravel()[:, None] ** 2)
-                    ).tocsr()[k]
-                )
-                sigma_w_inv += np.diag(1 / prior_sigma ** 2)
-                # Fit the flux - 1
-                B = A3[k].T.dot(
-                    (
-                        (flux_binned[seg_mask].ravel() - 1)
-                        / flux_err_binned[seg_mask].ravel() ** 2
-                    )[k]
-                )
-                B += prior_mu / (prior_sigma ** 2)
-                time_model_w = np.linalg.solve(sigma_w_inv, B)
-                res = flux_binned[seg_mask] - A3.dot(time_model_w).reshape(
-                    flux_binned[seg_mask].shape
+        # get uncontaminated pixel norm-flux
+        flux, flux_err = np.array(
+            [
+                np.array(
+                    [
+                        self.source_mask.multiply(self.flux[idx]).data,
+                        self.source_mask.multiply(self.flux_err[idx]).data,
+                    ]
                 )
-                res = np.ma.masked_array(res, (~k).reshape(flux_binned[seg_mask].shape))
-                bad_targets = sigma_clip(res, sigma=5).mask
-                bad_targets = (
-                    np.ones(flux_binned[seg_mask].shape, bool) & bad_targets.any(axis=0)
-                ).ravel()
-                # k &= ~sigma_clip(flux_binned.ravel() - A3.dot(time_model_w)).mask
-                k &= ~bad_targets
-            #  we save the weights and pixel mask of each segement for later use
-            seg_time_model_w.append(time_model_w)
-            pix_mask_k.append(k)
-
-        self.seg_splits = splits
-        self.time_model_w = np.asarray(seg_time_model_w)
-        self._time_masked = np.asarray(pix_mask_k, dtype=object)
+                for idx in range(self.nt)
+            ]
+        ).transpose((1, 0, 2))
+        flux_norm = flux / np.nanmean(flux, axis=0)
+        flux_err_norm = flux_err / np.nanmean(flux, axis=0)
+
+        # bindown flux arrays
+        flux_binned_raw = P3.bin_func(flux)
+        flux_binned = P3.bin_func(flux_norm)
+        flux_err_binned = P3.bin_func(flux_err_norm)
```

Comment:
Make sure to use the `quad` keyword here otherwise it will not bin correctly!
```suggestion
        flux_err_binned = P3.bin_func(flux_err_norm, quad=True)
```